components/slot-fill/test/slot.js

```diff
@@ -96,4 +96,37 @@ describe( 'Slot', () => {
 
 		expect( element.find( 'Slot > div' ).html() ).toBe( '<div>2</div>' );
 	} );
+
+	it( 'should render in expected order', () => {
+		const element = mount(
+			<Provider>
+				<Slot name="egg" key="slot" />
+			</Provider>
+		);
+
+		element.setProps( {
+			children: [
+				<Slot name="egg" key="slot" />,
+				<Filler name="egg" key="first" text="first" />,
+				<Filler name="egg" key="second" text="second" />,
+			],
+		} );
+
+		element.setProps( {
+			children: [
+				<Slot name="egg" key="slot" />,
+				<Filler name="egg" key="second" text="second" />,
+			],
+		} );
+
+		element.setProps( {
+			children: [
+				<Slot name="egg" key="slot" />,
+				<Filler name="egg" key="first" text="first" />,
+				<Filler name="egg" key="second" text="second" />,
+			],
+		} );
```

Comment:
```js
element.setProps( {
	children: [
		<Slot name="egg" key="slot" />,
 		<Filler name="egg" key="second" text="second" />,
 		<Filler name="egg" key="first" text="first" />,
 	],
 } )
```
I tried the following instead of line 122 and it outputs `<div>secondfirst</div>`. Which means you can use still use `concatChildren`, because it doesn't care about the keys passed. However it is predictable because it resets occurences for every slot whenever fill are removed.

If I add the same code after line 128, it still renders `<div>firstsecond</div>` because occurrences are not modified.